csrc/preseg_passes/segment_inplace_update.cpp

```diff
@@ -0,0 +1,156 @@
+// we check for implict broadcasts if any aliased input already has a
+// broadcast dimension that is concretized later in the fusion. This ensures
+// that all write operations to the fusion inputs occur after the read
+// operations have completed. See Issue #2664: https://
+// github.com/NVIDIA/Fuser/issues/2664
+namespace {
+void insertSegmentSet(Fusion* fusion) {
+  std::vector<TensorView*> aliased_tvs;
+  std::vector<BroadcastOp*> bcast;
+
+  // Find all tensorviews which are used in inplace updates
+  for (TensorView* tv : fusion->allTvs()) {
+    if (fusion->getOutputAlias(tv).type == AllocationType::ReuseBuffer) {
+      aliased_tvs.push_back(tv);
+    }
+  }
+
+  // Return early if there is no inplace update
+  if (aliased_tvs.empty()) {
+    return;
+  }
+
+  // fusion->exprs() is a topologically sorted list. Filter out the broadcast
+  // ops from the list.
+  auto all_exprs = fusion->exprs();
+  auto all_bcast_ops = ir_utils::filterByType<BroadcastOp>(all_exprs);
+
+  // Traverse and store all direct/indirect consumer tensorviews of these
+  // broadcast nodes If the tensorview has been visited, return --> this means
+  // that we have already traversed that branch
+  std::unordered_set<TensorView*> visited_tvs;
+  for (auto bcast_op : all_bcast_ops) {
+    std::deque<TensorView*> tvs_to_visit;
+    tvs_to_visit.push_back(bcast_op->output(0)->as<TensorView>());
+    while (!tvs_to_visit.empty()) {
+      TensorView* current_tv = tvs_to_visit.front();
+      tvs_to_visit.pop_front();
+      if (visited_tvs.count(current_tv)) {
+        continue;
+      }
+      visited_tvs.insert(current_tv);
+      std::vector<Expr*> current_tv_uses = current_tv->uses();
+      for (Expr* use : current_tv_uses) {
+        for (auto output_tv :
+             ir_utils::filterByType<TensorView>(use->outputs())) {
+          tvs_to_visit.push_back(output_tv->as<TensorView>());
+        }
+      }
+    }
+  }
+
+  // Traverse and store the direct/indirect producer tensorviews of these
+  // broadcast nodes If that tensorview has been visited, return.
+  for (auto bcast_op : all_bcast_ops) {
+    std::deque<TensorView*> tvs_to_visit;
+    tvs_to_visit.push_back(bcast_op->input(0)->as<TensorView>());
+    while (!tvs_to_visit.empty()) {
+      TensorView* current_tv = tvs_to_visit.front();
+      tvs_to_visit.pop_front();
+      if (visited_tvs.count(current_tv)) {
+        continue;
+      }
+      visited_tvs.insert(current_tv);
+      auto definition = current_tv->definition();
+      if (definition != nullptr) {
+        for (auto input_tv :
+             ir_utils::filterByType<TensorView>(definition->inputs())) {
+          tvs_to_visit.push_back(input_tv->as<TensorView>());
+        }
+      }
+    }
+  }
+
+  // Use permissive IdModel graph to identify any concretized broadcast
+  // iterdomain in any aliased input.
+  auto id_model =
+      IdModel(fusion, /*build_graphs=*/false, /*allow_self_mapping=*/true);
```

Comment:
another naive question. What's the case with self_mapping here? i.e. is it expected to have self mapping in the graph or it's just hiding some issue that should be tracked and made aware by @naoyam ?

Reply:
This is not needed here, removed now.